File: java/benchmark/src/main/java/de/hhu/bsinfo/observatory/benchmark/MessagingLatencyOperation.java
package de.hhu.bsinfo.observatory.benchmark;

import de.hhu.bsinfo.observatory.benchmark.Connection.Mode;
import de.hhu.bsinfo.observatory.benchmark.result.Status;

public class MessagingLatencyOperation extends LatencyOperation {

    private final String OUTPUT_FILENAME = "Messaging Latency";

    MessagingLatencyOperation(Connection connection, Mode mode, int operationCount, int operationSize) {
        super(connection, mode, operationCount, operationSize);
    }

    @Override
    String getOutputFilename() {
        return OUTPUT_FILENAME;
    }

    @Override
    boolean needsFilledReceiveQueue() {
        return getMode() != Mode.SEND;
    }

    @Override
    Status warmUp(int operationCount) {
        if (getMode() == Mode.SEND) {
            for (int i = 0; i < operationCount; i++) {
                Status status = getConnection().sendSingleMessage();

                if (status != Status.OK) {
                    return status;
                }
            }

            return Status.OK;
        } else {
            return getConnection().receiveMultipleMessages(operationCount);
        }
    }

    @Override
    Status execute() {
        if (getMode() == Mode.SEND) {
            long startTime = System.nanoTime();

            for (int i = 0; i < getMeasurement().getOperationCount(); i++) {
                getMeasurement().startSingleMeasurement();
                Status status = getConnection().sendSingleMessage();
                getMeasurement().stopSingleMeasurement();

                if (status != Status.OK) {
                    return status;
                }
            }

            getMeasurement().finishMeasuring(System.nanoTime() - startTime);

            return Status.OK;
        } else {
            return getConnection().receiveMultipleMessages(getMeasurement().getOperationCount());
        }
    }
}

File: cpp/src/observatory/operation/MessagingLatencyOperation.cpp
#include "MessagingLatencyOperation.h"

namespace Observatory {

MessagingLatencyOperation::MessagingLatencyOperation(Benchmark *benchmark, Benchmark::Mode mode, uint32_t operationCount, uint32_t operationSize) :
        LatencyOperation(benchmark, mode, operationCount, operationSize) {}

const char *MessagingLatencyOperation::getClassName() const {
    return "MessagingLatencyOperation";
}

const char* MessagingLatencyOperation::getOutputFilename() const {
    return "Messaging Latency";
}

bool MessagingLatencyOperation::needsFilledReceiveQueue() const {
    return getMode() != Benchmark::Mode::SEND;
}

Status MessagingLatencyOperation::warmUp(uint32_t operationCount) {
    if(getMode() == Benchmark::Mode::SEND) {
        for(uint32_t i = 0; i < operationCount; i++) {
            Status status = getBenchmark().sendSingleMessage();

            if(status != Status::OK) {
                return status;
            }
        }

        return Status::OK;
    } else {
        return getBenchmark().receiveMultipleMessages(operationCount);
    }
}

Status MessagingLatencyOperation::execute() {
    if(getMode() == Benchmark::Mode::SEND) {
        auto startTime = std::chrono::high_resolution_clock::now();

        for(uint32_t i = 0; i < getMeasurement().getOperationCount(); i++) {
            getMeasurement().startSingleMeasurement();
            Status status = getBenchmark().sendSingleMessage();
            getMeasurement().stopSingleMeasurement();

            if(status != Status::OK) {
                return status;
            }
        }

        uint64_t time = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now() - startTime).count();
        getMeasurement().finishMeasuring(time);

        return Status::OK;
    } else {
        return getBenchmark().receiveMultipleMessages(getMeasurement().getOperationCount());
    }
}

}